yjit/src/codegen.rs

```diff
@@ -2759,33 +2737,44 @@ fn gen_setinstancevariable(
     // then just write out the IV write as a function call.
     // too-complex shapes can't use index access, so we use rb_ivar_get for them too.
     if !receiver_t_object || uses_custom_allocator || shape_too_complex || new_shape_too_complex || megamorphic {
-        asm_comment!(asm, "call rb_vm_setinstancevariable()");
-
-        let ic = jit.get_arg(1).as_u64(); // type IVC
-
         // The function could raise FrozenError.
         // Note that this modifies REG_SP, which is why we do it first
         jit_prepare_non_leaf_call(jit, asm);
 
         // Get the operands from the stack
         let val_opnd = asm.stack_opnd(0);
 
-        // Call rb_vm_setinstancevariable(iseq, obj, id, val, ic);
-        asm.ccall(
-            rb_vm_setinstancevariable as *const u8,
-            vec![
-                Opnd::const_ptr(jit.iseq as *const u8),
-                Opnd::mem(64, CFP, RUBY_OFFSET_CFP_SELF),
-                ivar_name.into(),
-                val_opnd,
-                Opnd::const_ptr(ic as *const u8),
-            ]
-        );
+        if let StackOpnd(index) = recv_opnd { // attr_writer
+            let recv = asm.stack_opnd(index as i32);
+            asm_comment!(asm, "call rb_vm_set_ivar_id()");
+            asm.ccall(
+                rb_vm_set_ivar_id as *const u8,
+                vec![
+                    recv,
+                    Opnd::UImm(ivar_name),
+                    val_opnd,
+                ],
+            );
+        } else { // setinstancevariable
+            asm_comment!(asm, "call rb_vm_setinstancevariable()");
+            asm.ccall(
+                rb_vm_setinstancevariable as *const u8,
+                vec![
+                    Opnd::const_ptr(jit.iseq as *const u8),
+                    Opnd::mem(64, CFP, RUBY_OFFSET_CFP_SELF),
+                    ivar_name.into(),
+                    val_opnd,
+                    Opnd::const_ptr(ic.unwrap() as *const u8),
+                ],
+            );
+        }
```

Comment:
It seems like a good idea to reuse code and it's nice to see a 2% speedup! 👍 

However, this PR introduces multiple if statements to use different operands and change the behavior, which seems a bit weird.

Just to make sure I understand, attr_writers do not have ICs? That's why you have to call a different function here?

Reply:
You're right. attr_writers don't have ICs we need here, so attr_writers can't use `rb_vm_setinstancevariable`. An attr_writer is a method call, so it has an `cd->cc` (method call cache), but it doesn't have an instance-variable cache `ic` (`IVC`), which is used by setinstancevariable.

Similarly, setinstancevariable can't use `rb_vm_set_ivar_id`. attr_writer fallbacks to a dynamic dispatch when the receiver is not `T_OBJECT`, but setinstancevariable wants to use this general case for non-`T_OBJECT` cases.

We could use `rb_ivar_set` for both cases, which supports non-`T_OBJECT` receivers as well, but I saw a slowdown when I tried that, so I kept this logic. But let me benchmark it again 🤔

Reply:
> We could use rb_ivar_set for both cases, which supports non-T_OBJECT receivers as well, but I saw a slowdown when I tried that, so I kept this logic. But let me benchmark it again 🤔

It does seem to give a slowdown (these benchmarks consistently give a 0.99x change). It seems to hurt the performance if you handle both cases using `rb_ivar_set`. Another thing we could try is to replace `rb_vm_setinstancevariable` with `rb_ivar_set`, but it doesn't remove this `if` branch, so it doesn't simplify code.

```
before: ruby 3.4.0dev (2024-02-22T01:26:29Z yjit-attrset f27198fac3) +YJIT [x86_64-linux]
after: ruby 3.4.0dev (2024-02-22T17:12:46Z yjit-attrset 2b73548263) +YJIT [x86_64-linux]

----------  -----------  ----------  ----------  ----------  -------------  ------------
bench       before (ms)  stddev (%)  after (ms)  stddev (%)  after 1st itr  before/after
hexapdf     1862.5       3.0         1874.3      1.9         1.02           0.99
railsbench  2423.5       0.9         2436.3      0.8         0.98           0.99
----------  -----------  ----------  ----------  ----------  -------------  ------------
```

Reply:
> Another thing we could try is to replace rb_vm_setinstancevariable with rb_ivar_set, but it still leaves this if branch, so it doesn't simplify code.

I also tried this, but it also gives a slowdown (0.99x) on railsbench. So it seems more efficient to keep using `rb_vm_setinstancevariable` for setinstancevariable and `rb_vm_set_ivar_id` for attr_writer.

Reply:
Oh yeah, I wasn't suggesting that we use the less efficient function for both. The code is a bit harder to follow though with the special casing, but I understand why it is there.